src/autoload.rs

```diff
@@ -54,26 +96,103 @@ impl Autoload {
     /// After returning a path, the command is marked in-progress until the caller calls
     /// mark_autoload_finished() with the same command. Note this does not actually execute any
     /// code; it is the caller's responsibility to load the file.
-    pub fn resolve_command(&mut self, cmd: &wstr, env: &dyn Environment) -> Option<WString> {
+    pub fn resolve_command(&mut self, cmd: &wstr, env: &dyn Environment) -> Option<AutoloadPath> {
+        use crate::autoload::AutoloadPath;
+        use crate::wchar_ext::WExt;
+
+        let mut possible_path = None;
         if let Some(var) = env.get(self.env_var_name) {
-            self.resolve_command_impl(cmd, var.as_list())
+            match self.resolve_command_impl(cmd, var.as_list()) {
+                AutoloadResult::Path(path) => {
+                    crate::FLOGF!(autoload, "Loading from path with var: %ls", path);
+                    // HACK: Ignore generated_completions until we tried the embedded assets
+                    if path
+                        .find("/generated_completions/".chars().collect::<Vec<_>>())
+                        .is_some()
+                    {
+                        possible_path = Some(path);
+                    } else {
+                        return Some(AutoloadPath::Path(path));
+                    }
+                }
+                AutoloadResult::Loaded => return None,
+                AutoloadResult::Pending => return None,
+                AutoloadResult::None => (),
+            };
         } else {
-            self.resolve_command_impl(cmd, &[])
+            match self.resolve_command_impl(cmd, &[]) {
+                AutoloadResult::Path(path) => {
+                    crate::FLOGF!(autoload, "Loading from path with var: %ls", path);
+                    return Some(AutoloadPath::Path(path));
+                }
+                AutoloadResult::Loaded => return None,
+                AutoloadResult::Pending => return None,
+                AutoloadResult::None => (),
+            };
         }
+
+        // HACK: In cargo tests, this used to never load functions
+        // It will hang for reasons unrelated to this.
+        #[cfg(test)]
+        return None;
+
+        #[cfg(feature = "embed-data")]
+        {
+            let narrow = wcs2string(cmd);
+            let cmdstr = std::str::from_utf8(&narrow).ok()?;
```

Comment:
okay rust embed doesn't support non-UTF-8 paths. We can probably live with that.

Reply:
Yeah, we would have to name one of our functions with non-utf8, and I think we can just decide not to do that.